File: runtime-src/Monk/Runtime/Exec.hs
{-# LANGUAGE ForeignFunctionInterface #-}
{-# LANGUAGE OverloadedStrings #-}

-- | Bounded executable invocation and source-aware kernel exec diagnostics.
-- There is deliberately no ENOEXEC fallback to a shell source interpreter.
module Monk.Runtime.Exec (dispatchExecSite, executionFailure) where

import Control.Exception (IOException, catch)
import Control.Monad (when)
import Data.ByteString qualified as B
import Foreign.C.Error (Errno, eACCES, eNOENT, eNOEXEC, eNOTDIR, errnoToIOError)
import Foreign.C.String (CString)
import Foreign.C.Types (CInt (..))
import GHC.Foreign qualified as Foreign
import GHC.IO.Encoding (getFileSystemEncoding)
import Monk.Runtime.Descriptors (initialDescriptorOpen, nativeErrorMessage)
import Monk.Runtime.Protocol (protocolFailure)
import Monk.Runtime.Spawn (execProcess)
import System.Exit (ExitCode (ExitFailure), exitWith)
import System.IO (hFlush, stderr)
import System.Posix.Env.ByteString (getEnvironment)
import System.Posix.Types (Fd (..))

foreign import ccall unsafe "monk_directory_at" directoryAt :: CInt -> CString -> IO CInt

executionFailure :: Maybe Fd -> B.ByteString -> B.ByteString -> B.ByteString -> Errno -> IO (Int, B.ByteString)
executionFailure cwd origin line command failure = do
  directory <- if failure == eACCES then (/= 0) <$> B.useAsCString command (directoryAt (maybe (-1) (\(Fd fd) -> fd) cwd)) else pure False
  message <-
    if directory
      then pure "Is a directory"
      else
        if failure == eNOEXEC
          then pure "cannot execute binary file: Exec format error"
          else
            if not (B.elem 47 command) && failure `elem` [eNOENT, eNOTDIR]
              then pure "command not found"
              else nativeErrorMessage (errnoToIOError "execute" failure Nothing Nothing)
  let code = if failure == eNOENT || (not (B.elem 47 command) && failure == eNOTDIR) then 127 else 126
  pure (code, origin <> ": line " <> line <> ": " <> command <> ": " <> message <> "\n")

dispatchExecSite :: [String] -> IO ()
dispatchExecSite (origin : line : command : arguments) = do
  source <- nativeBytes origin
  row <- nativeBytes line
  executable <- nativeBytes command
  values <- mapM nativeBytes arguments
  environment <- filter (\(name, _) -> name `notElem` ["MONK_LAUNCH_ORIGINAL", "MONK_LAUNCH_WRAPPER"]) <$> getEnvironment
  failure <- execProcess environment executable values
  (code, diagnostic) <- executionFailure Nothing source row executable failure
  visible <- initialDescriptorOpen (2 :: Int)
  when visible ((B.hPut stderr diagnostic >> hFlush stderr) `catch` (\(_ :: IOException) -> pure ()))
  exitWith (ExitFailure code)
dispatchExecSite _ = protocolFailure "exec-site needs origin, line and executable"

nativeBytes :: FilePath -> IO B.ByteString
nativeBytes path = getFileSystemEncoding >>= \encoding -> Foreign.withCString encoding path B.packCString

File: runtime/src/exec.rs
//! Bounded execution with source-aware diagnostics; never an ENOEXEC interpreter.
use crate::native;
use std::{
    io,
    os::{fd::AsFd, unix::ffi::OsStrExt},
};

pub fn execution_failure(
    cwd: Option<native::BorrowedDirectory<'_>>,
    origin: &[u8],
    line: &[u8],
    command: &[u8],
    failure: &io::Error,
) -> (i32, Vec<u8>) {
    let error = failure.raw_os_error().unwrap_or(libc::EBADF);
    let directory = error == libc::EACCES && {
        let path = std::ffi::OsStr::from_bytes(command);
        let stat = match cwd {
            Some(fd) => rustix::fs::statat(fd, path, rustix::fs::AtFlags::empty()),
            None => rustix::fs::stat(path),
        };
        stat.is_ok_and(|status| {
            rustix::fs::FileType::from_raw_mode(status.st_mode) == rustix::fs::FileType::Directory
        })
    };
    let message = if directory {
        b"Is a directory".to_vec()
    } else if error == libc::ENOEXEC {
        b"cannot execute binary file: Exec format error".to_vec()
    } else if !command.contains(&b'/') && matches!(error, libc::ENOENT | libc::ENOTDIR) {
        b"command not found".to_vec()
    } else {
        native::native_error_message(failure)
    };
    let code = if error == libc::ENOENT || (!command.contains(&b'/') && error == libc::ENOTDIR) {
        127
    } else {
        126
    };
    (
        code,
        [
            origin, b": line ", line, b": ", command, b": ", &message, b"\n",
        ]
        .concat(),
    )
}
pub fn dispatch(args: &[Vec<u8>]) -> io::Result<i32> {
    let [origin, line, command, rest @ ..] = args else {
        return Err(io::Error::new(
            io::ErrorKind::InvalidInput,
            "exec-site needs origin, line and executable",
        ));
    };
    let environment = native::environment()
        .into_iter()
        .filter(|(key, _)| key != b"MONK_LAUNCH_ORIGINAL" && key != b"MONK_LAUNCH_WRAPPER")
        .collect::<Vec<_>>();
    let failure = native::exec_process(&environment, command, rest);
    let (code, diagnostic) = execution_failure(None, origin, line, command, &failure);
    if native::initial_descriptor_open(2)
        && let Ok(fd) = native::inherited_fd(2)
    {
        let _ = native::write_all(fd.as_fd(), &diagnostic);
    }
    Ok(code)
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn command_not_found_and_format_errors() {
        assert_eq!(
            execution_failure(
                None,
                b"source",
                b"4",
                b"absent",
                &io::Error::from_raw_os_error(libc::ENOENT)
            ),
            (127, b"source: line 4: absent: command not found\n".to_vec())
        );
        assert_eq!(
            execution_failure(
                None,
                b"source",
                b"4",
                b"./bad",
                &io::Error::from_raw_os_error(libc::ENOEXEC)
            ),
            (
                126,
                b"source: line 4: ./bad: cannot execute binary file: Exec format error\n".to_vec()
            )
        );
    }
}
